torch/serialization.py

```diff
@@ -303,45 +265,101 @@ def _check_container_source(container_type, source_file, original_source):
+        data = saved_id[1:]
+
+        if typename == 'module':
             # Ignore containers that don't have any sources saved
-            if all(saved_id[1:]):
-                _check_container_source(*saved_id)
-            return saved_id[0]
-        return deserialized_objects[int(saved_id)]
-
-    with closing(tarfile.open(fileobj=f, mode='r:', format=tarfile.PAX_FORMAT)) as tar, \
-            mkdtemp() as tmpdir:
-
-        tar.extract('storages', path=tmpdir)
-        with open(os.path.join(tmpdir, 'storages'), 'rb', 0) as f:
-            num_storages = pickle_module.load(f)
-            for i in range(num_storages):
-                args = pickle_module.load(f)
-                key, location, storage_type = args
-                obj = storage_type._new_with_file(f)
-                obj = restore_location(obj, location)
-                deserialized_objects[key] = obj
-
-            storage_views = pickle_module.load(f)
-            for target_cdata, root_cdata, offset, size in storage_views:
-                root = deserialized_objects[root_cdata]
-                deserialized_objects[target_cdata] = root[offset:offset + size]
-
-        tar.extract('tensors', path=tmpdir)
-        with open(os.path.join(tmpdir, 'tensors'), 'rb', 0) as f:
-            num_tensors = pickle_module.load(f)
-            for i in range(num_tensors):
-                args = pickle_module.load(f)
-                key, storage_id, original_tensor_type = args
-                storage = deserialized_objects[storage_id]
-                tensor_type = storage_to_tensor_type(storage)
-                tensor = tensor_type._new_with_metadata_file(f, storage)
-                deserialized_objects[key] = tensor
-
-        pickle_file = tar.extractfile('pickle')
-        unpickler = pickle_module.Unpickler(pickle_file)
-        unpickler.persistent_load = persistent_load
-        result = unpickler.load()
-        return result
+            if all(data[1:]):
+                _check_container_source(*data)
+            return data[0]
+        elif typename == 'storage':
+            data_type, key, location, size, is_view, offset, view_size = data
+            if key not in deserialized_storages:
+                deserialized_storages[key] = restore_location(
+                    data_type(size), location)
+            data = deserialized_storages[key]
+            if is_view:
+                data = data[offset:offset + view_size]
+            return data
+        else:
+            raise RuntimeError("Unknown saved id type: %s" % saved_id[0])
+
+    # try the legacy loader first, which only works if f is a tarfile
+    try:
+        return legacy_load(f)
+    except tarfile.TarError:
+        pass
+
+    f.seek(0)
+    magic_number = pickle_module.load(f)
+    if magic_number != MAGIC_NUMBER:
+        raise RuntimeError("Invalid magic number; corrupt file?")
+    protocol_version = pickle_module.load(f)
+    if protocol_version != PROTOCOL_VERSION:
```

Comment:
Err, maybe I misunderstand what this value means. Does it refer to the pickle protocol version? (because it is never used in that way). I thought that this is a version flag for backwards compatibility, i.e. so if we change the serialization scheme we can check the protocol version and decide which to use (which would have been a great way to ensure backwards compatibility here except that the protocol version was stored inside a nested tar file 👎 )